flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/StateChangeLogger.java

```diff
@@ -0,0 +1,122 @@
+ */
+
+package org.apache.flink.state.changelog;
+
+import org.apache.flink.core.memory.DataOutputViewStreamWrapper;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.function.BiConsumerWithException;
+import org.apache.flink.util.function.ThrowingConsumer;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Map;
+
+interface StateChangeLogger<State, Namespace> {
+    static <Namespace, State, StateElement> Iterator<StateElement> loggingIterator(
+            @Nullable Iterator<StateElement> iterator,
+            StateChangeLogger<State, Namespace> changeLogger,
+            BiConsumerWithException<StateElement, DataOutputViewStreamWrapper, IOException>
+                    changeWriter,
+            Namespace ns) {
+        if (iterator == null) {
+            return null;
+        }
+        return new Iterator<StateElement>() {
+
+            @Nullable private StateElement lastReturned;
+
+            @Override
+            public boolean hasNext() {
+                return iterator.hasNext();
+            }
+
+            @Override
+            public StateElement next() {
+                return lastReturned = iterator.next();
+            }
+
+            @Override
+            public void remove() {
+                try {
+                    changeLogger.valueElementRemoved(
+                            out -> changeWriter.accept(lastReturned, out), ns);
+                } catch (IOException e) {
+                    ExceptionUtils.rethrow(e);
+                }
+                iterator.remove();
+            }
+        };
+    }
+
+    static <Namespace, State, StateElement> Iterable<StateElement> loggingIterable(
+            @Nullable Iterable<StateElement> iterable,
+            KvStateChangeLogger<State, Namespace> changeLogger,
+            BiConsumerWithException<StateElement, DataOutputViewStreamWrapper, IOException>
+                    changeWriter,
+            Namespace ns) {
+        if (iterable == null) {
+            return null;
+        }
+        return () -> loggingIterator(iterable.iterator(), changeLogger, changeWriter, ns);
+    }
+
+    static <UK, UV, State, Namespace> Map.Entry<UK, UV> loggingMapEntry(
+            Map.Entry<UK, UV> entry,
+            KvStateChangeLogger<State, Namespace> changeLogger,
+            BiConsumerWithException<Map.Entry<UK, UV>, DataOutputViewStreamWrapper, IOException>
+                    changeWriter,
+            Namespace ns) {
+        return new Map.Entry<UK, UV>() {
+            @Override
+            public UK getKey() {
+                return entry.getKey();
+            }
+
+            @Override
+            public UV getValue() {
+                return entry.getValue();
+            }
+
+            @Override
+            public UV setValue(UV value) {
+                try {
+                    changeLogger.valueElementChanged(out -> changeWriter.accept(entry, out), ns);
```

Comment:
Do you need to log the value as well?

If you look at the how `changeWriter` is defined in ChangelogMapState, it only logs the key.

```
private final BiConsumerWithException<
                    Map.Entry<UK, UV>, DataOutputViewStreamWrapper, IOException>
            changeWriter = (entry, out) -> serializeKey(entry.getKey(), out);
```

Reply:
Right!